Add Bazel support.

jin opened the pull request:
This PR adds support for Bazel BUILD and WORKSPACE file generation, gated by the `generateBazelFiles` configuration option.

This has been used to generate a wide array of projects at https://github.com/jin/android-projects. 

What's in this PR:

- `generateBazelFiles` configuration option
- Generation of a top level WORKSPACE file, with external dependency on https://github.com/bazelbuild/gmaven_rules/ to download AARs from Google Maven.
- Generation of a BUILD.bazel for each Android and Java module, with dependencies wired up. If it's a top level Android module, an `android_binary` target is generated. Otherwise, it's either an `android_library` or `java_library` target.

What's not in this PR:

- Testing support (this is next on the list)
- Kotlin support

jin committed:
Generate BUILD.bazel files, but with build.gradle contents

jin committed:
Generate android_library, android_binary and java_library targets, and make them buildable with Bazel

jin committed:
Only generate test dependencies (junit, espresso, runner) if generateTests is true

jin committed:
Add support for gmaven_rules to download AARs from Google Maven

jin committed:
Add multidex support

jin committed:
Rebase and fixed merge conflicts

jin committed:
Update GMAVEN_TAG

jin committed:
Minimize diff

jin committed:
Merge branch 'master' into bazel

jin committed:
Cleanup

jin committed:
Add "generateBazelFiles" to configuration.

Set it to true to generate a Bazel WORKSPACE file at the top level, and a BUILD
file for each module.

jin committed:
Add license header, and remove generateBazelFiles from default configuration.

jin committed:
Fixed invalid JSON

jin committed:
Add tests for AndroidBuildBazelBlueprint

jin committed:
Add tests for Bazel in ConfigPojoToBuildSystemConfigConverter

jin committed:
Add information about Bazel support to the sample configs and README.

jin committed:
Add tests for ModuleBuildBazelGenerator

jin committed:
Add tests for AndroidModuleBuildBazelGenerator

jin committed:
Add tests for ModuleBuildBazelBlueprint

jin committed:
Fixed JSON for ConfigFull

jin committed:
Only generate Bazel files for Android modules if generateBazelFiles is set

NikitaKozlov commented on aspoet/src/main/kotlin/com/google/androidstudiopoet/generators/SourceModuleGenerator.kt:
```diff
@@ -56,15 +77,15 @@ class SourceModuleGenerator(private val moduleBuildGradleGenerator: ModuleBuildG
             projectBlueprint.moduleBlueprints.asReversed().forEach { blueprint ->
                 val job = launch {
-                    writeModule(blueprint)
+                    writeModule(blueprint, projectBlueprint.generateBazelFiles)
```
Please, make this boolean a part of the `ModuleBlueprint` and `AndroidModuleBlueprint`, then `buildBazelBlueprint` can be optional in both of them and if it is present, then the `BUILD.bazel` file is generated by respective generator

NikitaKozlov commented on aspoet/src/main/kotlin/com/google/androidstudiopoet/generators/SourceModuleGenerator.kt:
```diff
@@ -45,6 +46,26 @@ class SourceModuleGenerator(private val moduleBuildGradleGenerator: ModuleBuildG
         fileWriter.delete(projectBlueprint.projectRoot)
         fileWriter.mkdir(projectBlueprint.projectRoot)
 
+        if (projectBlueprint.generateBazelFiles) {
```
Please extract this into `BazelWorkspaceBlueprint` and `BazelWorkspaceGenerator`, to keep this file small (although it should be even smaller :) ).

NikitaKozlov commented on aspoet/src/main/kotlin/com/google/androidstudiopoet/models/ModuleBuildBazelBlueprint.kt:
```diff
@@ -0,0 +1,35 @@
+        moduleRoot: String
+) : ModuleBuildSpecificationBlueprint {
+
+    override val moduleName = moduleRoot.split("/").last()
```
I'm not sure if this will work on Windows... You can pass it in the constructor. It is called from `ModuleBlueprint`, so you can use `AbstractModuleBlueprint.name` property there. It has the value you are looking for..

NikitaKozlov commented on aspoet/src/main/kotlin/com/google/androidstudiopoet/models/AndroidBuildGradleBlueprint.kt:
```diff
@@ -23,17 +23,17 @@ import com.google.androidstudiopoet.input.PluginConfig
-                                  pluginConfigs: List<PluginConfig>?) {
-    val plugins: Set<String> = createSetOfPlugins(pluginConfigs)
+                                  pluginConfigs: List<PluginConfig>?) : AndroidModuleBuildSpecificationBlueprint {
+    override val plugins: Set<String> = createSetOfPlugins(pluginConfigs)
```
I'm not sure if this will work on Windows... You can pass it in the constructor. It is called from `ModuleBlueprint`, so you can use `AbstractModuleBlueprint.name` property there. It has the value you are looking for..

NikitaKozlov commented on aspoet/src/main/kotlin/com/google/androidstudiopoet/generators/android_modules/AndroidModuleBuildBazelGenerator.kt:
```diff
@@ -0,0 +1,55 @@
+class AndroidModuleBuildBazelGenerator(val fileWriter: FileWriter) {
+    fun generate(bazelBlueprint: AndroidBuildBazelBlueprint) {
+
+        val deps: Set<String> = bazelBlueprint.dependencies.mapNotNull {
```
You can use simply `map` here instead on `mapNotNull`

NikitaKozlov commented on aspoet/src/main/kotlin/com/google/androidstudiopoet/generators/ModuleBuildBazelGenerator.kt:
```diff
@@ -0,0 +1,45 @@
+
+class ModuleBuildBazelGenerator(private val fileWriter: FileWriter) {
+    fun generate(blueprint: ModuleBuildSpecificationBlueprint) {
+        val deps: Set<String> = blueprint.dependencies.mapNotNull {
```
You can use `map` here instead of `mapNotNull`

NikitaKozlov commented on aspoet/src/main/kotlin/com/google/androidstudiopoet/generators/android_modules/AndroidModuleBuildBazelGenerator.kt:
```diff
@@ -0,0 +1,55 @@
+import com.google.androidstudiopoet.models.*
+import com.google.androidstudiopoet.writers.FileWriter
+
+class AndroidModuleBuildBazelGenerator(val fileWriter: FileWriter) {
```
What do you think, would it make sense to have an analog for `GradleLang.kt.` for Bazel? Then support and modification of Bezel-related files might be easier.

jin replied:
I think this PR is large enough, so I'll create the language abstractions in a follow up PR.

jin replied:
https://github.com/android/android-studio-poet/pull/135

NikitaKozlov reviewed (commented):
Thank you very much for this PR. It is great to have Bazel on board.

jin committed:
Use map instead of mapNotNull

jin committed:
Extract generateBazelFiles configuration into AndroidModuleBlueprint and
ModuleBlueprint.

If it's not null and true, generate Bazel BUILD files accordingly.

jin committed:
Extract Bazel WORKSPACE file generation semantics into BazelWorkspaceBlueprint
and BazelWorkspaceGenerator.

jin committed:
Add tests for BazelWorkspaceGenerator and BazelWorkspaceBlueprint

jin committed:
Move Gradle-only properties from {Android}ModuleBuildSpecificationBlueprint to
{Android}ModuleBuildGradleBlueprint.

Use AbstractModuleBlueprint#name as the target name for Bazel targets.

jin committed:
Remove ModuleBuildSpecificationBlueprint supertype since build.gradle and
BUILD.bazel share very little.

jin committed:
Removed redundant AndroidModuleBuildSpecificationBlueprint

jin committed:
Reduce line diffs.

NikitaKozlov commented on aspoet/src/main/kotlin/com/google/androidstudiopoet/generators/android_modules/AndroidModuleGenerator.kt:
```diff
@@ -43,6 +44,10 @@ class AndroidModuleGenerator(private val resourcesGenerator: ResourcesGenerator,
         blueprint.activityBlueprints.forEach({ activityGenerator.generate(it) })
         manifestGenerator.generate(blueprint)
+
+        if (blueprint.generateBazelFiles != null && blueprint.generateBazelFiles) {
```
TBH my idea was bit different. I thought that if `blueprint.buildBazelBlueprint` is optional then this code can look something like 
```
blueprint.buildBazelBlueprint?.let { buildBazelGenerator.generate(it) }
```
, then generator doesn't need to know about `blueprint.generateBazelFiles`

jin replied:
https://github.com/android/android-studio-poet/pull/134
